### PR TITLE
Upgrade to LLVM 16 & C++17, and bundle libc++ on Linux

### PR DESCRIPTION
**Pull request by directhex**
This includes building against, linking, and shipping our own libc++ in lieu of system libstdc++, for LLVM builds. This is because LLVM 16 has a hard requirement on C++17 - our existing code assumes C++11

### CODE REVIEW

**Comment by directhex** [comment]
Browser stuff is a mess, currently somewhat blocked on https://github.com/dotnet/runtime/pull/86806 since the Ubuntu images just don't align with what I need (e.g. only having GCC on the Ubuntu WebAssembly image, and apparently using that to build the cross compiler on the WebAssembly leg)

I'm going to look at LLVM 16 integration testing in some other repos for now

File: src/mono/mono/tools/offsets-tool/offsets-tool.py [hunk 1]
```diff
@@ -372,8 +372,11 @@ def gen (self):
 			if type.size == -1:
 				continue
 			f.write ("DECL_SIZE2(%s,%s)\n" % (type.name, type.size))
+			done_fields = []
 			for field in type.fields:
```
Comment (vargaz) [inline src/mono/mono/tools/offsets-tool/offsets-tool.py:376]: This should be a dictionary, i.e. ```done_fields = {}``` and later ```done_fields [field.name] = field.name```

File: src/mono/mono/tools/offsets-tool/offsets-tool.py [hunk 2]
```diff
@@ -383,8 +386,11 @@ def gen (self):
...
```
Comment (vargaz) [inline src/mono/mono/tools/offsets-tool/offsets-tool.py:390]: Ditto.

**Comment by directhex** [comment]
Extra platforms Android failures are an infra issue - cbl-mariner-2.0-android image lacks an x64 crossrootfs, so cannot produce a cross-compiler

**Comment by directhex** [comment]
`runtime (Build linux-x64 Release AllSubsets_Mono_LLVMAot_RuntimeTests llvmaot)` is fixed by https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/898

**Comment by directhex** [comment]
If I'm correct, all the OSX NativeAOT failures are fixed by https://github.com/dotnet/llvm-project/pull/437

**Comment by directhex** [comment]
> > requirement on C++17 - our existing code assumes C++11
> 
> Just curious, is this because 17 isn't back compatible?

There's a whole bunch of includes used in LLVM 16+ that aren't available in pre-C++17, as such LLVM 16+ itself won't compile against a system `libstdc++` which lacks those headers. It became mandatory in this LLVM commit: https://github.com/llvm/llvm-project/commit/b1356504e63ae821cccf1e051a0d2526bdfef2b0

We do not have a sufficiently modern `libstdc++` in our Mariner images (since the crossrootfs is Ubuntu 16.04 which predates the definition of C++17), so our workaround is to build LLVM's `libc++` standard library and ship it alongside the runtime, as needed. We want to only link one C++ library, which has to be `libc++` for anything that also links LLVM or LLVM-based things. `libc++` is much stricter than `libstdc++` in its "this header is ONLY for a given standards version" ways, so various places where we've casually used C++14 features have just worked with `libstdc++`, but are hard failures with `libc++`.

The only viable alternatives to this route are:

a) Never update LLVM past 15, and cancel all plans for Microsoft builds of emsdk/wasi-sdk, which are a prerequisite for native stack windows-arm64 support for wasm/wasi targets
b) Bump Mariner's crosrootfs target & drop some Linux support (currently glibc 2.23 and libstdc++ 5.3.1, which covers e.g. Amazon Linux 2)
c) Build and ship our own `libstdc++` (GCC) instead of `libc++`

This wasn't an issue before the move to Mariner, because our CentOS 7 images used a post-release version of `libstdc++` via "devtoolsets"

File: src/mono/mono.proj [hunk 3]
```diff
@@ -55,6 +55,8 @@
     <MonoWasmThreads Condition="'$(MonoWasmBuildVariant)' == 'singlethread'">false</MonoWasmThreads>
     <MonoWasmThreads Condition="'$(MonoWasmBuildVariant)' == 'multithread'">true</MonoWasmThreads>
     <MonoEnableAssertMessages Condition="'$(MonoEnableAssertMessages)' == '' and '$(TargetsWasi)' == 'true'">true</MonoEnableAssertMessages>
+    <MonoLLVMMajorVersion>$(runtimelinuxx64MicrosoftNETCoreRuntimeMonoLLVMSdkVersion.Substring(0, $(runtimelinuxx64MicrosoftNETCoreRuntimeMonoLLVMSdkVersion.IndexOf('.'))))</MonoLLVMMajorVersion>
+    <MonoUseLibCxx Condition="'$(MonoLLVMMajorVersion)' >= '16.0'">true</MonoUseLibCxx>
```
Comment (akoeplinger) [inline src/mono/mono.proj:59]: I'm surprised this works, according to https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-conditions?view=vs-2022 you need to escape `>` to `&gt;`

Comment (directhex) [reply src/mono/mono.proj:59]: It's not clear to me from the docs whether you need to escape `>=` to `&gt;=`, or just `>` to `>=`. The fact that the former works implies it's not required

**Comment by filipnavara** [comment]
> Shrink the min Dwarf warning limit.

Sorry. I plead guilty.

**Comment by directhex** [comment]
> > Shrink the min Dwarf warning limit.
> 
> Sorry. I plead guilty.

Plz stop fixing bugs, those were load-bearing bugs

**Review by jkotas — APPROVED** [review]
:shipit:

**Review by vargaz — APPROVED** [review]

**Comment by jkotas** [comment]
I am wondering whether it would be better to link `libc++.so` and `libc++abi.so.1` statically into libobjwriter.so, so there is no risk of it colliding with system libraries that may depend on C++ runtime.

**Comment by agocke** [comment]
Agreed, statically linking libc++ would be fine with me.

Also, the warning shrink looks good to me, that may also have been impacted by https://github.com/dotnet/llvm-project/pull/444

**Comment by filipnavara** [comment]
> Also, the warning shrink looks good to me, that may also have been impacted by [dotnet/llvm-project#444](https://github.com/dotnet/llvm-project/pull/444)

Didn't mean to steal credit from you. 😅 Anyway, we are (back) at zero warnings with llvm-dwarfdump.

**Comment by agocke** [comment]
Nope, no worries, you've done tons in this space and deserve the credit. I was just pointing out that, if that change hadn't yet been integrated, it might cause us to lower the count once again. I have one more change that will lower the count again on Linux (for some reason Linux still seems to have errors, but Mac doesn't).

**Comment by filipnavara** [comment]
> for some reason Linux still seems to have errors, but Mac doesn't

They ship different dwarfdump. macOS ships the LLVM one, which reports less errors but supports proprietary Apple sections (which were later replaced by DWARF5 standard).

**Comment by directhex** [comment]
If someone wants to run experiments on static linking vs dynamic, they're welcome to - though the place to start that work is in the llvm-project repo where objwriter is generated. In my testing, via use of `rpath`, dynamic linking is working at least well enough for CI.

I don't think static linking is a realistic option for Mono - the bloat on 20-30 clang/llvm binaries would push us past the Nupkg size limit again

**Comment by jkotas** [comment]
This introduced build break in Android extra-platforms Mono legs:
 
https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/328385/logs/223

```
2023-07-03T16:59:17.6848371Z   CMake Error at /usr/share/cmake-3.21/Modules/CMakeTestCXXCompiler.cmake:62 (message):
2023-07-03T16:59:17.6856690Z     The C++ compiler
2023-07-03T16:59:17.6856932Z   
2023-07-03T16:59:17.6857259Z       "/usr/bin/c++"
2023-07-03T16:59:17.6857646Z   
2023-07-03T16:59:17.6857894Z     is not able to compile a simple test program.
2023-07-03T16:59:17.6858043Z   
2023-07-03T16:59:17.6858208Z     It fails with the following output:
2023-07-03T16:59:17.6858340Z   
2023-07-03T16:59:17.6858549Z       Change Dir: /__w/1/s/artifacts/obj/mono/android.arm64.Release/cross/CMakeFiles/CMakeTmp
2023-07-03T16:59:17.6858948Z       
2023-07-03T16:59:17.6859436Z       Run Build Command(s):/usr/bin/make -f Makefile cmTC_e98b0/fast && /usr/bin/make  -f CMakeFiles/cmTC_e98b0.dir/build.make CMakeFiles/cmTC_e98b0.dir/build
2023-07-03T16:59:17.6859987Z       make[1]: Entering directory '/__w/1/s/artifacts/obj/mono/android.arm64.Release/cross/CMakeFiles/CMakeTmp'
2023-07-03T16:59:17.6860286Z       Building CXX object CMakeFiles/cmTC_e98b0.dir/testCXXCompiler.cxx.o
2023-07-03T16:59:17.6861105Z       /usr/bin/c++   -I/__w/1/s/artifacts/obj/mono/android.arm64.Release/llvm//x64/include/c++/v1 -L/__w/1/s/artifacts/obj/mono/android.arm64.Release/llvm//x64/lib -stdlib=libc++  -o CMakeFiles/cmTC_e98b0.dir/testCXXCompiler.cxx.o -c /__w/1/s/artifacts/obj/mono/android.arm64.Release/cross/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
2023-07-03T16:59:17.6862084Z       c++: error: unrecognized command-line option ‘-stdlib=libc++’
2023-07-03T16:59:17.6862397Z       make[1]: *** [CMakeFiles/cmTC_e98b0.dir/build.make:78: CMakeFiles/cmTC_e98b0.dir/testCXXCompiler.cxx.o] Error 1
2023-07-03T16:59:17.6862993Z       make[1]: Leaving directory '/__w/1/s/artifacts/obj/mono/android.arm64.Release/cross/CMakeFiles/CMakeTmp'
2023-07-03T16:59:17.6863231Z       make: *** [Makefile:127: cmTC_e98b0/fast] Error 2
2023-07-03T16:59:17.
```

**Pull request merged**